// @redis/search/dist/lib/commands/SUGGET_WITHSCORES_WITHPAYLOADS.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const generic_transformers_1 = require("@redis/client/dist/lib/commands/generic-transformers");
const SUGGET_1 = __importDefault(require("./SUGGET"));
exports.default = {
    FIRST_KEY_INDEX: SUGGET_1.default.FIRST_KEY_INDEX,
    IS_READ_ONLY: SUGGET_1.default.IS_READ_ONLY,
    transformArguments(...args) {
        const transformedArguments = SUGGET_1.default.transformArguments(...args);
        transformedArguments.push('WITHSCORES', 'WITHPAYLOADS');
        return transformedArguments;
    },
    transformReply: {
        2(reply) {
            if ((0, generic_transformers_1.isNullReply)(reply))
                return null;
            const transformedReply = new Array(reply.length / 3);
            let replyIndex = 0, arrIndex = 0;
            while (replyIndex < reply.length) {
                transformedReply[arrIndex++] = {
                    suggestion: reply[replyIndex++],
                    score: Number(reply[replyIndex++]),
                    payload: reply[replyIndex++]
                };
            }
            return transformedReply;
        },
        3(reply) {
            if ((0, generic_transformers_1.isNullReply)(reply))
                return null;
            const transformedReply = new Array(reply.length / 3);
            let replyIndex = 0, arrIndex = 0;
            while (replyIndex < reply.length) {
                transformedReply[arrIndex++] = {
                    suggestion: reply[replyIndex++],
                    score: reply[replyIndex++],
                    payload: reply[replyIndex++]
                };
            }
            return transformedReply;
        }
    }
};
