// @redis/time-series/dist/lib/commands/MRANGE.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformMRangeArguments = exports.pushGroupByArgument = exports.TIME_SERIES_REDUCERS = void 0;
const RANGE_1 = require("./RANGE");
const MGET_1 = require("./MGET");
exports.TIME_SERIES_REDUCERS = {
    AVG: 'AVG',
    SUM: 'SUM',
    MIN: 'MIN',
    MAX: 'MAX',
    RANGE: 'RANGE',
    COUNT: 'COUNT',
    STD_P: 'STD.P',
    STD_S: 'STD.S',
    VAR_P: 'VAR.P',
    VAR_S: 'VAR.S'
};
function pushGroupByArgument(args, groupBy) {
    if (groupBy) {
        args.push('GROUPBY', groupBy.label, 'REDUCE', groupBy.reducer);
    }
    return args;
}
exports.pushGroupByArgument = pushGroupByArgument;
function transformMRangeArguments(command, fromTimestamp, toTimestamp, filter, options) {
    let args = (0, RANGE_1.pushRangeArguments)([command], fromTimestamp, toTimestamp, options);
    args = (0, MGET_1.pushFilterArgument)(args, filter);
    return pushGroupByArgument(args, options?.GROUPBY);
}
exports.transformMRangeArguments = transformMRangeArguments;
exports.default = {
    FIRST_KEY_INDEX: undefined,
    IS_READ_ONLY: true,
    transformArguments: transformMRangeArguments.bind(undefined, 'TS.MRANGE'),
    // TODO
    transformReply: undefined
};
