// @redis/graph/dist/lib/commands/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const CONFIG_GET_1 = __importDefault(require("./CONFIG_GET"));
const CONFIG_SET_1 = __importDefault(require("./CONFIG_SET"));
;
const DELETE_1 = __importDefault(require("./DELETE"));
const EXPLAIN_1 = __importDefault(require("./EXPLAIN"));
const LIST_1 = __importDefault(require("./LIST"));
const PROFILE_1 = __importDefault(require("./PROFILE"));
const QUERY_1 = __importDefault(require("./QUERY"));
const RO_QUERY_1 = __importDefault(require("./RO_QUERY"));
const SLOWLOG_1 = __importDefault(require("./SLOWLOG"));
exports.default = {
    CONFIG_GET: CONFIG_GET_1.default,
    configGet: CONFIG_GET_1.default,
    CONFIG_SET: CONFIG_SET_1.default,
    configSet: CONFIG_SET_1.default,
    DELETE: DELETE_1.default,
    delete: DELETE_1.default,
    EXPLAIN: EXPLAIN_1.default,
    explain: EXPLAIN_1.default,
    LIST: LIST_1.default,
    list: LIST_1.default,
    PROFILE: PROFILE_1.default,
    profile: PROFILE_1.default,
    QUERY: QUERY_1.default,
    query: QUERY_1.default,
    RO_QUERY: RO_QUERY_1.default,
    roQuery: RO_QUERY_1.default,
    SLOWLOG: SLOWLOG_1.default,
    slowLog: SLOWLOG_1.default
};
