// @redis/client/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCluster = exports.createClientPool = exports.createClient = exports.defineScript = exports.VerbatimString = exports.RESP_TYPES = void 0;
var decoder_1 = require("./lib/RESP/decoder");
Object.defineProperty(exports, "RESP_TYPES", { enumerable: true, get: function () { return decoder_1.RESP_TYPES; } });
var verbatim_string_1 = require("./lib/RESP/verbatim-string");
Object.defineProperty(exports, "VerbatimString", { enumerable: true, get: function () { return verbatim_string_1.VerbatimString; } });
var lua_script_1 = require("./lib/lua-script");
Object.defineProperty(exports, "defineScript", { enumerable: true, get: function () { return lua_script_1.defineScript; } });
// export * from './lib/errors';
const client_1 = __importDefault(require("./lib/client"));
exports.createClient = client_1.default.create;
const pool_1 = require("./lib/client/pool");
exports.createClientPool = pool_1.RedisClientPool.create;
const cluster_1 = __importDefault(require("./lib/cluster"));
exports.createCluster = cluster_1.default.create;
// export { GeoReplyWith } from './lib/commands/generic-transformers';
// export { SetOptions } from './lib/commands/SET';
// export { RedisFlushModes } from './lib/commands/FLUSHALL';
