// @redis/client/dist/lib/commands/ACL_LOG.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    FIRST_KEY_INDEX: undefined,
    IS_READ_ONLY: true,
    transformArguments(count) {
        const args = ['ACL', 'LOG'];
        if (count !== undefined) {
            args.push(count.toString());
        }
        return args;
    },
    transformReply: {
        2: (reply) => {
            return reply.map(item => {
                const inferred = item;
                return {
                    count: inferred[1],
                    reason: inferred[3],
                    context: inferred[5],
                    object: inferred[7],
                    username: inferred[9],
                    'age-seconds': Number(inferred[11]),
                    'client-info': inferred[13],
                    'entry-id': inferred[15],
                    'timestamp-created': inferred[17],
                    'timestamp-last-updated': inferred[19]
                };
            });
        },
        3: undefined
    }
};
