// @redis/client/dist/lib/commands/BZPOPMAX.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformBZPopArguments = void 0;
const generic_transformers_1 = require("./generic-transformers");
function transformBZPopArguments(command, key, timeout) {
    const args = (0, generic_transformers_1.pushVariadicArguments)([command], key);
    args.push(timeout.toString());
    return args;
}
exports.transformBZPopArguments = transformBZPopArguments;
exports.default = {
    FIRST_KEY_INDEX: 1,
    IS_READ_ONLY: false,
    transformArguments(...args) {
        return transformBZPopArguments('BZPOPMAX', ...args);
    },
    transformReply: {
        2(reply) {
            return reply === null ? null : {
                key: reply[0],
                value: reply[1],
                score: Number(reply[2])
            };
        },
        3(reply) {
            return reply === null ? null : {
                key: reply[0],
                value: reply[1],
                score: reply[2]
            };
        }
    }
};
