// @redis/client/dist/lib/commands/CLIENT_KILL.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLIENT_KILL_FILTERS = void 0;
exports.CLIENT_KILL_FILTERS = {
    ADDRESS: 'ADDR',
    LOCAL_ADDRESS: 'LADDR',
    ID: 'ID',
    TYPE: 'TYPE',
    USER: 'USER',
    SKIP_ME: 'SKIPME'
};
exports.default = {
    FIRST_KEY_INDEX: undefined,
    IS_READ_ONLY: true,
    transformArguments(filters) {
        const args = ['CLIENT', 'KILL'];
        if (Array.isArray(filters)) {
            for (const filter of filters) {
                pushFilter(args, filter);
            }
        }
        else {
            pushFilter(args, filters);
        }
        return args;
    },
    transformReply: undefined
};
function pushFilter(args, filter) {
    if (filter === exports.CLIENT_KILL_FILTERS.SKIP_ME) {
        args.push('SKIPME');
        return;
    }
    args.push(filter.filter);
    switch (filter.filter) {
        case exports.CLIENT_KILL_FILTERS.ADDRESS:
            args.push(filter.address);
            break;
        case exports.CLIENT_KILL_FILTERS.LOCAL_ADDRESS:
            args.push(filter.localAddress);
            break;
        case exports.CLIENT_KILL_FILTERS.ID:
            args.push(typeof filter.id === 'number' ?
                filter.id.toString() :
                filter.id);
            break;
        case exports.CLIENT_KILL_FILTERS.TYPE:
            args.push(filter.type);
            break;
        case exports.CLIENT_KILL_FILTERS.USER:
            args.push(filter.username);
            break;
        case exports.CLIENT_KILL_FILTERS.SKIP_ME:
            args.push(filter.skipMe ? 'yes' : 'no');
            break;
    }
}
