// @redis/client/dist/lib/commands/GEOSEARCH.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pushGeoSearchOptions = exports.pushGeoSearchArguments = void 0;
function pushGeoSearchArguments(args, key, from, by, options) {
    args.push(key);
    if (typeof from === 'string' || from instanceof Buffer) {
        args.push('FROMMEMBER', from);
    }
    else {
        args.push('FROMLONLAT', from.longitude.toString(), from.latitude.toString());
    }
    if ('radius' in by) {
        args.push('BYRADIUS', by.radius.toString(), by.unit);
    }
    else {
        args.push('BYBOX', by.width.toString(), by.height.toString(), by.unit);
    }
    pushGeoSearchOptions(args, options);
    return args;
}
exports.pushGeoSearchArguments = pushGeoSearchArguments;
function pushGeoSearchOptions(args, options) {
    if (options?.SORT) {
        args.push(options.SORT);
    }
    if (options?.COUNT) {
        if (typeof options.COUNT === 'number') {
            args.push('COUNT', options.COUNT.toString());
        }
        else {
            args.push('COUNT', options.COUNT.value.toString());
            if (options.COUNT.ANY) {
                args.push('ANY');
            }
        }
    }
}
exports.pushGeoSearchOptions = pushGeoSearchOptions;
exports.default = {
    FIRST_KEY_INDEX: 1,
    IS_READ_ONLY: true,
    transformArguments(key, from, by, options) {
        return pushGeoSearchArguments(['GEOSEARCH'], key, from, by, options);
    },
    transformReply: undefined
};
