// @redis/client/dist/lib/commands/GEOSEARCH_WITH.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GEO_REPLY_WITH = void 0;
const GEOSEARCH_1 = __importDefault(require("./GEOSEARCH"));
exports.GEO_REPLY_WITH = {
    DISTANCE: 'WITHDIST',
    HASH: 'WITHHASH',
    COORDINATES: 'WITHCOORD'
};
exports.default = {
    FIRST_KEY_INDEX: GEOSEARCH_1.default.FIRST_KEY_INDEX,
    IS_READ_ONLY: GEOSEARCH_1.default.IS_READ_ONLY,
    transformArguments(key, from, by, replyWith, options) {
        const args = GEOSEARCH_1.default.transformArguments(key, from, by, options);
        args.push(...replyWith);
        args.preserve = replyWith;
        return args;
    },
    transformReply(reply, replyWith) {
        const replyWithSet = new Set(replyWith);
        let index = 0;
        const distanceIndex = replyWithSet.has(exports.GEO_REPLY_WITH.DISTANCE) && ++index, hashIndex = replyWithSet.has(exports.GEO_REPLY_WITH.HASH) && ++index, coordinatesIndex = replyWithSet.has(exports.GEO_REPLY_WITH.COORDINATES) && ++index;
        return reply.map(raw => {
            const unwrapped = raw;
            const item = {
                member: unwrapped[0]
            };
            if (distanceIndex) {
                item.distance = unwrapped[distanceIndex];
            }
            if (hashIndex) {
                item.hash = unwrapped[hashIndex];
            }
            if (coordinatesIndex) {
                const [longitude, latitude] = unwrapped[coordinatesIndex];
                item.coordinates = {
                    longitude,
                    latitude
                };
            }
            return item;
        });
    }
};
