// @redis/client/dist/lib/commands/ROLE.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    FIRST_KEY_INDEX: undefined,
    IS_READ_ONLY: true,
    transformArguments() {
        return ['ROLE'];
    },
    transformReply(reply) {
        switch (reply[0]) {
            case 'master': {
                const [role, replicationOffest, replicas] = reply;
                return {
                    role,
                    replicationOffest,
                    replicas: replicas.map(replica => {
                        const [host, port, replicationOffest] = replica;
                        return {
                            host,
                            port: Number(port),
                            replicationOffest: Number(replicationOffest)
                        };
                    })
                };
            }
            case 'slave': {
                const [role, masterHost, masterPort, state, dataReceived] = reply;
                return {
                    role,
                    master: {
                        host: masterHost,
                        port: masterPort
                    },
                    state,
                    dataReceived,
                };
            }
            case 'sentinel': {
                const [role, masterNames] = reply;
                return {
                    role,
                    masterNames
                };
            }
        }
    }
};
